### ISSUE TITLE
Pull multiple images together

### ISSUE THREAD

**Issue by arun-gupta** [docker/docker#16106]
There should be an ability to pull multiple Docker images together. Today it gives:

```
~> docker pull jboss/wildfly arungupta/wildfly-management arungupta/javaee7-hol mysql swarm arungupta/wildfly-mysql-javaee7
docker: "pull" requires 1 argument.
See 'docker pull --help'.

Usage:	docker pull [OPTIONS] NAME[:TAG|@DIGEST]

Pull an image or a repository from a registry
```

Here are version details:

```
~> docker info
Containers: 0
Images: 0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.0.9-boot2docker
Operating System: Boot2Docker 1.8.0 (TCL 6.3); master : 7f12e95 - Tue Aug 11 17:55:16 UTC 2015
CPUs: 1
Total Memory: 996.2 MiB
Name: docker-java
ID: 2KB4:XZRD:QK3Y:7TRS:XN6R:JJRU:SULN:TCWN:BM3A:RYB4:OC4D:M6RQ
Debug mode (server): true
File Descriptors: 9
Goroutines: 16
System Time: 2015-09-06T15:39:55.733499441Z
EventsListeners: 0
Init SHA1: 
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker
Username: arungupta
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
```

**Comment by jfrazelle** [comment]
The fastly cache has been reset

**Comment by arun-gupta** [comment]
I'm not sure I understand how this helps with the issue :)

Can you please explain more?

**Comment by jfrazelle** [comment]
Wow wrong email responded sorry

**Comment by Pensu** [comment]
+1.

We should have the same provision for push too. docker push also requires only 1 argument.

**Comment by aneshas** [comment]
#dibs

**Comment by thaJeztah** [comment]
@Pensu there's an issue for push: https://github.com/docker/docker/issues/7336, and a pull-request; https://github.com/docker/docker/pull/15624, but it was put on hold because there are some issues with push that need to be resolved first


@aneshas thanks for working on this, it would be good to check with @stevvooe if he expects issues with this (we had some issues with parallel pulls that have only recently been resolved)

ping @stevvooe do you foresee problems implementing this, or is it good to go?

**Comment by stevvooe** [comment]
@thaJeztah We do not yet want to extend the syntax of push/pull before deciding on the method for specifying remotes. Most of the performance benefits gained from such a change can be achieved without changing the UX. This also complicates the implementation of push/pull.

I'd like to see the push/pull code fully refactored with download management, a remote specification and full optimizations before we accept any more sugar in these commands.

**Comment by thaJeztah** [comment]
@aneshas really sorry, but I think it's best to not start working on this (see the comment above). If you want to work on something, and need help finding an issue to work on, feel free to join the #docker-dev IRC channel, and ask for assistance.

(thanks @stevvooe, I expected this would be the case, thanks for confirming)

**Comment by aneshas** [comment]
@stevvooe @thaJeztah No problem I understand :D I'll just find another issue to work on :)

**Comment by tobegit3hub** [comment]
It seems that docker supports to pull multiple images at the same time after 1.10.

Should we close this?

**Comment by thaJeztah** [comment]
@tobegit3hub no this is not yet implemented; `docker pull imagea imageb images` does not work yet.

**Comment by arun-gupta** [comment]
@thaJeztah any updates on this?

**Comment by thaJeztah** [comment]
@arun-gupta I haven't heard about updates in this area, but possibly @stevvooe or @aaronlehmann knows more

**Comment by stevvooe** [comment]
We want to reserve this syntax to allow git-style remotes:

```
docker pull <remote> <image>
docker push <remove> <image>
```

If we allow multiple image pulls in a single command, we will not be able to replicate the familiar syntax.

The performance impact of individual push commands should be negligible versus doing it in a single command. If that is not the case, we should fix the performance impact.

**Comment by TimurKiyivinski** [comment]
Just a suggestion, what about more bash-like syntax?
`docker pull {imagea,imageb,imagec}`
That shouldn't conflict the git-style remotes

**Comment by aaronlehmann** [comment]
> Just a suggestion, what about more bash-like syntax?
> `docker pull {imagea,imageb,imagec}`
> That shouldn't conflict the git-style remotes

The problem with this is that it uses metacharacters which are normally interpreted by the shell. The argument would have to be quoted to prevent the shell from transforming it into something else. In theory it works, but I suspect it would be confusing.

**Comment by mafanasyev-tri** [comment]
Suggestion: Add `-m`/`--multiple` option, and require it for all multiple image pulls. This way, git-style syntax will still be reserved.

For me, the ability to pull multiple images increases parallelism. I have few images which have common ancestor and one big unique layer. Currently I run "pull" sequentially, and as a result, most of the time is spent downloading that one layer using a single thread. Multi-image pull should significantly increase performance by allowing multi-threaded downloads.

**Comment by stevvooe** [comment]
@mafanasyev-tri We might be able to do something simpler by having a `-t,--tag` flag that takes multiple values:

```
docker pull -t foo -t bar localhost:5000/myimage
docker push -t foo -t bar localhost:5000/myimage
```

The above would operate on the image `localhost:5000/myimage:latest`, pushing and pulling the set of tags associated with that image, in addition to the implicit or explicit tag.

I also think we should add syntax for a kind of "push as" syntax, as a kind of "remotes light":

```
docker push localhost:5000/myimage mylocalimage:latest
```

The above would push `mylocalimage:latest` to `localhost:5000/myimage:latest`, bypassing the need for a tagging process. The reason this can work is that when remotes are added, they can take precedence over image name in a backwards compatible way.

That would preserve the possibility of remote-style syntax (which we are adding to containerd ;) ) and solve the initial complaint. In addition, it solves the long standing complaint of having to tag an image before pushing to another registry.

Sorry this took so long to find a solid proposal on: we just have a lot more clarity on where we want this to go. We've been able to experiment with a few things in containerd to make this more straightforward.

**Comment by mafanasyev-tri** [comment]
Thanks for thinking about this! 

FWIW, multi-tag fetching would be useless for me (and it would not solve initial complaint either). We use tags for version number, and we have a number of images, one for each major components. So we have just the opposite situation -- one tag and multiple images.

Regarding "remotes light": I am not sure I follow. How would this command (from the first message in this port) look?
` docker pull jboss/wildfly arungupta/wildfly-management arungupta/javaee7-hol`

**Comment by stevvooe** [comment]
> FWIW, multi-tag fetching would be useless for me (and it would not solve initial complaint either). We use tags for version number, and we have a number of images, one for each major components. So we have just the opposite situation -- one tag and multiple images.

I am speaking of this particular issue.

"Multi-threaded downloads" won't necessarily speed up pulls in the way that you envision. The files are processed sequentially, since they are compressed tar archives, so one must retrieve the blocks in order. In the case you describe (which is different from the original issue report), docker is already operating as "optimally" as it can, without drastically changing the image format.

In general, despite the syntactic changes, the speedups wouldn't necessarily be realized. The only actual speedup to be gained here would be pushing or pulling a set of common tags from the same repository.

> Regarding "remotes light": I am not sure I follow. How would this command (from the first message in this post) look?
docker pull jboss/wildfly arungupta/wildfly-management arungupta/javaee7-hol

Pull would not work with "remotes light". The above is proposed to handle the multiple tags case (which I guess isn't actually the original complaint).

**Comment by mafanasyev-tri** [comment]
> "Multi-threaded downloads" won't necessarily speed up pulls in the way that you envision. The files are processed sequentially, since they are compressed tar archives, so one must retrieve the blocks in order. In the case you describe (which is different from the original issue report), docker is already operating as "optimally" as it can, without drastically changing the image format.

I understand that you can extract one image at a time, but I was hoping for the speedup due to multi-threaded network download. Here is a specific example from  our latest jenkins job (which starts on fresh VM and downloads new images):

* Pull image `foo:v2017-06-23`
    * [0 sec] start download, 9 layers in parallel
    * [10 sec] all but one layers have been downloaded and unpacked, _"pull" is jut downloading a single layer_
    * [26 sec]: last layer has finished downloading
    * [126 sec]: last layer has finished unpacking,  "docker pull" exits
* Pull image `bar:v2017-06-23`
    * [0 sec] start download, 5 layers in parallel
    * [7 sec] all but one layers have been downloaded and unpacked, _"pull" is just downloading a single layer file_
    * [37 sec]: last layer has finished downloading
    * [160 sec]: last layer has finished unpacking,  "docker pull" exits
* Pull image `bar:v2017-06-23` (timeline omitted)

As you can see, during each phase there is a a few tens of second when "pull" does not do any extraction at all, and just downloads the single image file. I was hoping that in multiple image pull mode, docker would do everything in parallel -- by the time `foo`'s layers will be unpacked, `bar`'s layers would be all downloaded and ready to go.

**Comment by cpuguy83** [comment]
@mafanasyev-tri This is only a CLI issue. You can initiate multiple pulls simultaneously, though everything will slow down at this point.
That's more stuff taking bandwidth, CPU, and disk IO.

**Comment by stevvooe** [comment]
@mafanasyev-tri In that vein, are you asking that it does something like this:

```
echo wordpress redis ubuntu postgres | xargs -P10 -n1 docker pull
```

The above will use 10 process to run `docker pull` commands. The docker engine should largely parallelize common resources that are being requested at the same time.

**Comment by J0WI** [comment]
There are already some global commands in `docker system`, so how about `docker system pull-all`?

**Comment by sergey-morenets** [comment]
Any updates ?

**Comment by numToStr** [comment]
I solved this with a zsh for loop.

```zsh
for img (node:lts-slim mongo nginx:alpine); do docker pull $img; done
```

**Comment by mafanasyev-tri** [comment]
@numToStr , this loop downloads the images serially, there is no parallelism there at all.
see https://github.com/moby/moby/issues/16106#issuecomment-310695631 for details.

**Comment by n1ru4l** [comment]
**Workaround with bash (serially 😭)**
```bash
images=(node:lts-slim mongo nginx:alpine)
for img in "${images[@]}"; do docker pull $img; done
```

**Comment by instwpr** [comment]
@numToStr, running commands in bg with `&` could help a bit

**Comment by 0x2b3bfa0** [comment]
Same as https://github.com/moby/moby/issues/16106#issuecomment-733743239 but parallelized; at least in my case, performance was about the same as running the commands serially, though.

```bash
for image in node mongo nginx; do
  docker pull $image &
done
wait
```
